File: main/library-fine/library_fine.rb
# methods for treating arrays as due or returned [day, month, year]
class Array
  def day
    self[0]
  end

  def month
    self[1]
  end

  def year
    self[2]
  end
end

def read_date
  date = gets.split(' ').map(&:to_i)
  raise 'date record needs 3 fields (day, month, year)' unless date.size == 3
  date
end

def fine(returned, due)
  if returned.year < due.year then 0
  elsif returned.year > due.year then 10_000
  elsif returned.month < due.month then 0
  elsif returned.month > due.month then (returned.month - due.month) * 500
  elsif returned.day <= due.day then 0
  else (returned.day - due.day) * 15
  end
end

returned = read_date
due = read_date

puts fine(returned, due)

File: main/library-fine/library-fine.cs
using System;

internal static class Solution {
    private static int Day(this int[] date)     => date[0];
    private static int Month(this int[] date)   => date[1];
    private static int Year(this int[] date)    => date[2];

    private static int[] ReadDate()
    {
        return Array.ConvertAll(Console.ReadLine().Split(' '), int.Parse);
    }

    private static int CalculateFine(int[] returned, int[] due)
    {
        if (returned.Year() < due.Year()) return 0;
        if (returned.Year() > due.Year()) return 10000;

        if (returned.Month() < due.Month()) return 0;
        if (returned.Month() > due.Month())
            return (returned.Month() - due.Month()) * 500;

        if (returned.Day() <= due.Day()) return 0;
        return (returned.Day() - due.Day()) * 15;
    }

    private static void Main()
    {
        var returned = ReadDate();
        var due = ReadDate();

        Console.WriteLine(CalculateFine(returned, due));
    }
}
